go/teams/proofs.go

```diff
@@ -4,15 +4,76 @@ import (
 	"github.com/keybase/client/go/protocol/keybase1"
 )
 
-// Collection of ordering constraints waiting to be verified.
-// TODO implement
-type proofSetT struct{}
+func newProofTerm(i keybase1.UserOrTeamID, s keybase1.SignatureMetadata) proofTerm {
+	return proofTerm{leafID: i, sigMeta: s}
+}
+
+type proofTerm struct {
+	leafID  keybase1.UserOrTeamID
+	sigMeta keybase1.SignatureMetadata
+}
+
+type proofTermBookends struct {
+	left  proofTerm
+	right *proofTerm
+}
+
+type proof struct {
+	a proofTerm
+	b proofTerm
+}
+
+type proofIndex struct {
+	a keybase1.UserOrTeamID
+	b keybase1.UserOrTeamID
+}
+
+func (t proofTerm) seqno() keybase1.Seqno { return t.sigMeta.SigChainLocation.Seqno }
+
+func (t proofTerm) lessThanOrEqual(u proofTerm) bool {
+	return t.seqno() <= u.seqno()
+}
+
+func (t proofTerm) max(u proofTerm) proofTerm {
+	if t.lessThanOrEqual(u) {
+		return u
+	}
+	return t
+}
+
+func (t proofTerm) min(u proofTerm) proofTerm {
+	if t.lessThanOrEqual(u) {
+		return t
+	}
+	return u
+}
+
+func newProofIndex(a keybase1.UserOrTeamID, b keybase1.UserOrTeamID) proofIndex {
+	if string(a) < string(b) {
+		return proofIndex{a, b}
+	}
+	return proofIndex{b, a}
+}
+
+type proofSetT struct {
+	proofs map[proofIndex][]proof
+}
 
 func newProofSet() *proofSetT {
-	return &proofSetT{}
+	return &proofSetT{make(map[proofIndex][]proof)}
 }
 
-func (p *proofSetT) HappensBefore(a keybase1.SignatureMetadata, b keybase1.SignatureMetadata) *proofSetT {
-	// TODO!
+func (p *proofSetT) AddNeededHappensBeforeProof(a proofTerm, b proofTerm) *proofSetT {
```

Comment:
Cool! Suggested comment "Add a constraint that a must happen before b. If there is an existing assertion on the same pair of chains, and the new assertion is tighter, replace the existing one."